### ISSUE TITLE
Add option to delete caches to both kompile and ktest

### ISSUE THREAD

**Issue by cos** [kframework/k-legacy#697]
It is great that we now have kompilation caching, but, as expected, we get exceptions when kompiling after changing the code for KAST classes. This may not be only problematic scenario. We should have an option to `kompile` without using caches for both `kompile` and `ktest`. 

It could be something like `kompile --clean` and `ktest --clean`. Or maybe `--clean-slate`.

@radumereuta, what do you think?

**Comment by radumereuta** [comment]
Actually I found cases where it would be problematic to users too.
For example if you change the klabel of a production, the current implementation won't notice the modification. That's why I'm actually thinking of making the cache an option. Actually, an experimental one. Because this is exactly what it is, an experimental option.

**Comment by cos** [comment]
If we make it an option, it may take longer for it to stabilize as most users won't touch it. I would say we should leave it the default, and slowly learn what types of exceptions are thrown, catch them, and instruct the user to `kompile` with `--clean` when necessary.

**Comment by cos** [comment]
In the meantime, I'm getting an exception that I suspect is from caching. Are caches saved in any other place than `*-kompiled`? I remember something about a user directory.

**Comment by radumereuta** [comment]
Making it an option is the only thing that I am willing to do to this piece of code. I need to work on the parser, and I don't have time for anything else. If you want to take responsibility for it, I am willing to code review.

The only place where where we store caching information for rules is in *-kompiled/defx-cache.bin
Delete that file, and you are good to go again.

**Comment by cos** [comment]
Ok. I assigned myself to keep track of it. May bug you at some point if I need help.

**Comment by grosu** [comment]
I tend to agree with Radu here, we do not want to have beginners stare hopelessly at subtle errors.  The expert users get used with it.

Hm, I wonder if it wouldn't be a good idea to have a configuration file for K, where one can set all the options one likes, even the language name.  Then one can simply run

```
kompile
```

Grigore

Sent from phone

-------- Original message --------
From: Radu Mereuta
Date:07/03/2014 19:22 (GMT-06:00)
To: kframework/k
Subject: Re: [k] Add option to delete caches to both kompile and ktest (#697)

Making it an option is the only thing that I am willing to do to this piece of code. I need to work on the parser, and I don't have time for anything else. If you want to take responsibility for it, I am willing to code review.

The only place where where we store caching information for rules is in *-kompiled/defx-cache.bin
Delete that file, and you are good to go again.

**Comment by radumereuta** [comment]
Usually people use `make` for that.

**Comment by cos** [comment]
A common practice is to have a .something (.k for us) directory in the user's home with configurations specific to a the something tool. We could a kompile-options configuration file in there.

**Comment by cos** [comment]
It could also be a good place to store caches for shared (`include`) modules, once we have that ability.

**Comment by grosu** [comment]
I often erase .k when things get crazy, just to be sure.  For example, that's my clean option.  So it could be risky for users to store their configuration file there.  But true, make is a good option for that.  No rush to do it, we have way more urgent things to do.

Grigore

Sent from phone

-------- Original message --------
From: Cosmin Radoi
Date:07/03/2014 19:35 (GMT-06:00)
To: kframework/k
Cc: "Rosu, Grigore"
Subject: Re: [k] Add option to delete caches to both kompile and ktest (#697)

It could also be a good place to store caches for common modules, once we have that ability.

**Comment by cos** [comment]
Does k use a `.k` directory in the user's home directory (i.e., in `~`; e.g., `/Users/cos`) now? I think you're referring to the `.k` in the `k` directory.

**Comment by cos** [comment]
Regarding `make`, stretching the history of build systems a bit...
1. you can do anything by writing system code in C (or Java)
2.  `bash` is a better (i.e. shorter, easier to maintain, etc.) way of doing what can be done with system code in C
3. `make` was invented to put order into `bash` build scripts
4. `ant` replaced `make` as a cleaner solution for Java programs
5. `maven` replaces `ant` and adds dependency management, `ivy` adds dependency management on top of `ant`
6. `gradle` and `sbt` are more succinct replacements for `maven` -- they get rid of xml

Now, I have two questions:
- if `make` enters the mix, why does `k` use all the first four layers for its build and execution scripts? Was this a deliberate decision or did it just happen?
- why is `make` the right solution for a `--clean` sort of thing? As for what people use, they use any of the six layers for that.

**Comment by radumereuta** [comment]
I don't think there is a particular reason. It is just people preference and what they are familiar with.
I chose `ant` because I could easily generate it from eclipse, and Stratego and SDF already used `ant` in their build system. It was just more simple to adapt their scripts for our needs.

**Comment by dwightguth** [comment]
Oncall should implement this in the new parser since it just requires hooking in a single boolean value to the constructor of Kompile.java.
